c2rust-analyze/src/rewrite/expr/mir_op.rs

```diff
@@ -801,6 +958,58 @@ where
             return Ok(());
         }
 
+        if from.option && from.own != to.own {
+            // Downgrade ownership before unwrapping the `Option` when possible.  This can avoid
+            // moving/consuming the input.  For example, if the `from` type is `Option<Box<T>>` and
+            // `to` is `&mut T`, we start by calling `p.as_mut().as_deref()`, which gives
+            // `Option<&mut T>` without consuming `p`.
+            if !from.own.is_copy() {
+                match to.own {
+                    Ownership::Raw | Ownership::Imm => {
+                        (self.emit)(RewriteKind::OptionDowngrade {
+                            mutbl: false,
+                            deref: true,
+                        });
+                        from.own = Ownership::Imm;
+                    }
+                    Ownership::RawMut | Ownership::Cell | Ownership::Mut => {
+                        (self.emit)(RewriteKind::OptionDowngrade {
+                            mutbl: true,
+                            deref: true,
+                        });
+                        from.own = Ownership::Mut;
+                    }
+                    _ => {
+                        // Remaining cases are unsupported.
+                    }
+                }
+            }
+        }
+
+        let mut in_option_map = false;
+        if from.option && !to.option {
+            // Unwrap first, then perform remaining casts.
+            (self.emit)(RewriteKind::OptionUnwrap);
+            from.option = false;
+        } else if from.option && to.option {
+            eprintln!("try_build_cast_desc_desc: emit OptionMapBegin");
```

Comment:
Just a general comment: the analyzer should use a logger like `log` with e.g. `env_logger` as the back end, so all this debugging output can be silenced (`2>/dev/null` does technically work, but it's less configurable).

We don't need to fix all the existing lines in this PR, but maybe we could switch these? Or alternatively a follow-up PR that switches everything.

Reply:
I switched out the println calls that were added in this PR.